### PR TITLE
Added support for rizin fuzzing

### PR DESCRIPTION
**Pull request by m4drat**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -58,6 +58,7 @@ external repositories.
 * xlnt
 * ydb
 * zlib
+* rizin
```
Comment (SweetVishnya) [inline README.md:61]: alphabetical order

File: projects/rizin/sydr-fuzz.toml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+[sydr]
+args = "-l debug"
+target = "/rizin-fuzzing/sydr-plain/bin/rz-fuzz @@"
+symaddr = 25
```
Comment (SweetVishnya) [inline projects/rizin/sydr-fuzz.toml:20]: This is now set by default

File: projects/rizin/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+RUN apt-get update && \
+    apt-get -y install autoconf automake libtool ninja-build
+
+ENV PATH="${PATH}:/root/.local/bin"
+
+# Install meson
+RUN pip3 install --user meson==0.63.0
+
+# Clone target and corpus from github.
+RUN mkdir rizin-fuzzing && cd rizin-fuzzing && git clone https://github.com/rizinorg/rizin
```
Comment (SweetVishnya) [inline projects/rizin/Dockerfile:16]: We may clone rizin in root /rizin

File: projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	RzBinDwarfLineInfo *li = rz_bin_dwarf_parse_line(bf, info, RZ_BIN_DWARF_LINE_INFO_MASK_LINES);
+
+	rz_bin_dwarf_line_info_free(li);
+	rz_bin_dwarf_debug_info_free(info);
+	rz_bin_dwarf_debug_abbrev_free(da);
+	rz_buf_free(buff);
+	rz_bin_free(bin);
+
+	return 0;
+}
```
Comment (SweetVishnya) [inline projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c:49]: Add newline

File: projects/rizin/sydr-fuzz.toml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+[aflplusplus]
+args = "-t 4000+ -i /rizin-fuzzing/corpus"
+target = "/rizin-fuzzing/afl-asan/bin/rz-fuzz"
+jobs = 14
```
Comment (SweetVishnya) [inline projects/rizin/sydr-fuzz.toml:25]: It is better to set less jobs, so, one can run it on small pc

File: projects/rizin/sydr-fuzz.toml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# Copyright 2021 ISP RAS
```
Comment (SweetVishnya) [inline projects/rizin/sydr-fuzz.toml:1]: 2022

File: projects/rizin/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,36 @@
+FROM sweetvishnya/ubuntu20.04-sydr-fuzz
```
Comment (SweetVishnya) [inline projects/rizin/Dockerfile:1]: Add copyright

File: projects/rizin/build.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# Apply patches to support fuzzing
```
Comment (SweetVishnya) [inline projects/rizin/build.sh:1]: Add copyright

File: projects/rizin/rz-fuzz/main-impl.c [hunk 9]
```diff
@@ -0,0 +1,27 @@
+#include <stdint.h>
```
Comment (SweetVishnya) [inline projects/rizin/rz-fuzz/main-impl.c:1]: Add copyright

File: projects/rizin/rz-fuzz/meson.build [hunk 10]
```diff
@@ -0,0 +1,43 @@
+if get_option('fuzz_mode') == 'afl'
```
Comment (SweetVishnya) [inline projects/rizin/rz-fuzz/meson.build:1]: Add copyright

File: projects/rizin/rz-fuzz/rz-fuzz-afl.c [hunk 11]
```diff
@@ -0,0 +1,34 @@
+#include <stdio.h>
```
Comment (SweetVishnya) [inline projects/rizin/rz-fuzz/rz-fuzz-afl.c:1]: Add copyrigth

File: projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c [hunk 12]
```diff
@@ -0,0 +1,49 @@
+#include "rz_bin.h"
```
Comment (SweetVishnya) [inline projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c:1]: Add copyright

File: projects/rizin/sydr-fuzz.toml [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+[aflplusplus]
+args = "-t 4000+ -i /rizin-fuzzing/corpus"
+target = "/rizin-fuzzing/afl-asan/bin/rz-fuzz"
+jobs = 14
+rss_limit_mb = 16000
+
+[cov]
+# target = "/rizin-fuzzing/coverage-plain/bin/rz-fuzz @@"
+# Collect coverage for main binary
```
Comment (SweetVishnya) [inline projects/rizin/sydr-fuzz.toml:30]: Remove commented code

File: projects/rizin/sydr-fuzz.toml [hunk 14]
```diff
@@ -0,0 +1,31 @@
+# Copyright 2021 ISP RAS
+#
```
Comment (SweetVishnya) [inline projects/rizin/sydr-fuzz.toml:2]: Rename file to sydr-fuzz-afl++.toml

File: projects/rizin/sydr-support-fuzzing.diff [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
++  subdir('rz-fuzz')
+   subdir('rz-sign')
+   subdir('rz-ax')
+ endif
+diff --git a/meson_options.txt b/meson_options.txt
+index 4c565ccb3c..b5a9456d7f 100644
+--- a/meson_options.txt
++++ b/meson_options.txt
+@@ -1,3 +1,5 @@
++option('fuzz_mode', type: 'combo', choices: ['afl', 'libfuzzer', 'sydr', 'none'], value: 'none', description: 'Specifies fuzzing configuration to build (if any)')
```
Comment (SweetVishnya) [inline projects/rizin/sydr-support-fuzzing.diff:28]: Can you add .toml for libFuzzer + Sydr also?

File: projects/rizin/rz-fuzz/meson.build [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    dependencies: [
+      rz_util_dep,
+      rz_main_dep,
+      rz_socket_dep,
+      rz_cons_dep
+    ],
+    install: true,
+    install_rpath: rpath_exe,
+    implicit_include_directories: false,
+    cpp_args: '-fsanitize=fuzzer',
```
Comment (SweetVishnya) [inline projects/rizin/rz-fuzz/meson.build:26]: address,undefined

File: projects/rizin/build.sh [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+# Build sydr binary
+CURR_BUILD=sydr-plain
+mkdir /$CURR_BUILD
+CC=clang meson -Dfuzz_mode=sydr -Denable_tests=false -Denable_rz_test=false --default-library=static -Dstatic_runtime=false --buildtype=debugoptimized --prefix=/rizin-fuzzing/$CURR_BUILD build-$CURR_BUILD ; ninja -C build-$CURR_BUILD && ninja -C build-$CURR_BUILD install
+
+# Build with AFL++ instrumentation + ASAN enabled
+CURR_BUILD=afl-asan
+mkdir /$CURR_BUILD
+CC=afl-clang-lto CFLAGS="-fsanitize=address" LDFLAGS="-fsanitize=address" meson -Dfuzz_mode=afl -Denable_tests=false -Denable_rz_test=false --default-library=static -Dstatic_runtime=false --buildtype=debugoptimized --prefix=/rizin-fuzzing/$CURR_BUILD build-$CURR_BUILD ; ninja -C build-$CURR_BUILD && ninja -C build-$CURR_BUILD install
```
Comment (SweetVishnya) [inline projects/rizin/build.sh:12]: Add undefined sanitizer

File: projects/rizin/build.sh [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
```
Comment (SweetVishnya) [inline projects/rizin/build.sh:13]: Can we also build rizin for libFuzzer?

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (m4drat) [reply README.md]: Fixed

File: projects/rizin/sydr-fuzz.toml (hunk 2 above)
Comment (m4drat) [reply projects/rizin/sydr-fuzz.toml]: Fixed

File: projects/rizin/Dockerfile (hunk 3 above)
Comment (m4drat) [reply projects/rizin/Dockerfile:32]: I think it's better to have separate directory for all fuzzing-related things (binaries, source code, etc)

File: projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c (hunk 4 above)
Comment (m4drat) [reply projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c]: Fixed

File: projects/rizin/sydr-fuzz.toml (hunk 6 above)
Comment (m4drat) [reply projects/rizin/sydr-fuzz.toml]: Fixed

File: projects/rizin/Dockerfile (hunk 7 above)
Comment (m4drat) [reply projects/rizin/Dockerfile:17]: Fixed

File: projects/rizin/build.sh (hunk 8 above)
Comment (m4drat) [reply projects/rizin/build.sh:17]: Fixed

File: projects/rizin/rz-fuzz/main-impl.c (hunk 9 above)
Comment (m4drat) [reply projects/rizin/rz-fuzz/main-impl.c:17]: Fixed

File: projects/rizin/rz-fuzz/meson.build (hunk 10 above)
Comment (m4drat) [reply projects/rizin/rz-fuzz/meson.build:17]: Fixed

File: projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c (hunk 12 above)
Comment (m4drat) [reply projects/rizin/rz-fuzz/rz-fuzz-libfuzzer.c]: Fixed

File: projects/rizin/sydr-fuzz.toml (hunk 13 above)
Comment (m4drat) [reply projects/rizin/sydr-fuzz.toml]: Fixed

File: projects/rizin/sydr-fuzz.toml (hunk 14 above)
Comment (m4drat) [reply projects/rizin/sydr-fuzz.toml]: Fixed

File: projects/rizin/sydr-support-fuzzing.diff (hunk 15 above)
Comment (m4drat) [reply projects/rizin/sydr-support-fuzzing.diff:28]: Added .toml for libfuzzer

File: projects/rizin/build.sh (hunk 18 above)
Comment (m4drat) [reply projects/rizin/build.sh:34]: Yep, added new build target

File: projects/rizin/build.sh (hunk 17 above)
Comment (m4drat) [reply projects/rizin/build.sh]: Fixed

File: projects/rizin/sydr-fuzz-libfuzzer.toml [hunk 19]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2022 ISP RAS
```
Comment (SweetVishnya) [inline projects/rizin/sydr-fuzz-libfuzzer.toml:1]: Rename file to sydr-fuzz.toml

File: projects/rizin/build.sh [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+# Build with AFL++ instrumentation + ASAN enabled
+CURR_BUILD=afl-asan
+mkdir /$CURR_BUILD
+CC=afl-clang-lto CFLAGS="-fsanitize=address,undefined" LDFLAGS="-fsanitize=address,undefined" meson -Dfuzz_mode=afl -Denable_tests=false -Denable_rz_test=false --default-library=static -Dstatic_runtime=false --buildtype=debugoptimized --prefix=/rizin-fuzzing/$CURR_BUILD build-$CURR_BUILD ; ninja -C build-$CURR_BUILD && ninja -C build-$CURR_BUILD install
+
+# Build with libfuzzer
+CURR_BUILD=libfuzzer-asan
+mkdir /$CURR_BUILD
+CC=clang CFLAGS="-fsanitize=fuzzer-no-link,address" LDFLAGS="-fsanitize=fuzzer-no-link,undefined" meson -Dfuzz_mode=libfuzzer -Denable_tests=false -Denable_rz_test=false --default-library=static -Dstatic_runtime=false --buildtype=debugoptimized --prefix=/rizin-fuzzing/$CURR_BUILD build-$CURR_BUILD ; ninja -C build-$CURR_BUILD && ninja -C build-$CURR_BUILD install
```
Comment (SweetVishnya) [inline projects/rizin/build.sh:33]: Add address,undefined to both CFLAGS and LDFLAGS

File: projects/rizin/README.md [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+
+Unzip Sydr (`sydr.zip`) in `projects/rizin` directory:
+
+    $ unzip sydr.zip
+
+Run docker:
+
+    $ sudo docker run --privileged --network host -v /etc/localtime:/etc/localtime:ro --rm -it -v $PWD:/fuzz oss-sydr-fuzz-rizin /bin/bash
+
+Run hybrid fuzzing:
```
Comment (SweetVishnya) [inline projects/rizin/README.md:19]: Add docs for both afl++ and libFuzzer

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

**Pull request merged**